### ISSUE TITLE
Proposal: JSDoc only on exported nodes

### ISSUE THREAD

**Issue by orta** [eslint/eslint#7812]
**What rule do you want to change?**

`require-jsdoc` - [link](http://eslint.org/docs/rules/require-jsdoc)

Support making a rule so that the ESLinters for JSDocs are only applied on items which are going to leave the current source file ( and thus would be visible outside of the current scope. )

**Does this change cause the rule to produce more or fewer warnings?**

It would add an option to produce fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

There could be a `scope` attribute to require-jsdoc which allows you to define whether to apply the rules to all nodes. Or to only ones which will be available outside of it's file.

```json
{
    "require-jsdoc": ["error", {
       "scope": "all" || "exported",
        "require": {
            "FunctionDeclaration": true,
            "MethodDefinition": false,
            "ClassDeclaration": false,
            "ArrowFunctionExpression": false
        }
    }]
}

```

**Please provide some example code that this change will affect:**

for Modern: `import/export`

```js
export class MyThing {}
export function simpleFunction(one, two) {}
```

For the `require` style, something like this

```js
module.exports = class Runner extends EventEmitter {}

module.exports.jestChildProcessWithArgs = (workspace: ProjectWorkspace, args: Array<string>): ChildProcess => {}

```

So for some example code:

```js

var foo = (test) => {
    return test + 10;
}

/**
 * It returns 10
 */
export class Test{
    /**
    * returns the date
    */
    getDate(){}
}
```

**What does the rule currently do for this code?**

At the moment all JSDoc checks are applied wholesale throughout your source code. Meaning even internal functions, or test helper functions require comprehensive comments. So this would raise errors for `foo` but not `test`.

**What will the rule do after it's changed?**

With the scope as `exported` then the code above would not produce any errors.

**Comment by platinumazure** [comment]
Seems like a possible extension to `require-jsdoc`. Could you please put together a Rule Change Proposal?

**Comment by eslintbot** [comment]
Hi @orta, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by orta** [comment]
Ah yeah, good point, I have updated it as a change request proposal

**Comment by nzakas** [comment]
I don't think we currently check CommonJS exports, and if we do, I wouldn't want to group that into this option. 

If the option was just for ES6 exports, I have no problem with it. 

@orta would you be up for implementing this if accepted?

**Comment by orta** [comment]
If it was just for ES6 exports, yep. If it was all the different cases in commonJS, maybe not. Lots more edge cases there, and the syntax is new to me, only had to write ES6.

**Comment by vitorbal** [comment]
I would be in favor of this enhancement (for ES6 exports) :+1:

**Comment by kaicataldo** [comment]
Seems reasonable to me, though I'm not sold on the "scope" config option name.

**Comment by kaicataldo** [comment]
Thought about this some more and I actually don't have an suggestion for an alternative, so don't let my comment be a blocker.

**Comment by vitorbal** [comment]
@eslint/eslint-team We need one more +1 and a champion for this one. Any takers?

**Comment by BYK** [comment]
👍 (though not championing)

**Comment by nzakas** [comment]
@byk please add your 👍 to the issue description so we can keep track. 

For "scope" we could make it "exportedOnly" to change it into a Boolean option. We probably don't need a string unless we think there might be a third option.

**Comment by orta** [comment]
Yeah, I have no issue with the option `exportedOnly` does explain it better IMO

**Comment by platinumazure** [comment]
We have 5 :+1:s including me and BYK-- we just need a champion!

**Comment by nzakas** [comment]
Okay, I'll champion - I think this is a nice addition.

@orta feel free to submit a PR.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
